### PR TITLE
Beginner part 2 crud

### PR DESCRIPTION
**Pull request by ngavinsir**

### CODE REVIEW

File: app/cmd/root.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+	dbname   = "golangbeginner"
+)
+
+func init() {
+	dbConn := initDB()
+
+	paymentCodesRepository = postgresRepository.NewPaymentCodesRepository(dbConn)
+}
+
+func initDB() (db *sql.DB) {
```
Comment (owenyuwono) [inline app/cmd/root.go:29]: (nitpick)
reduce the usage of named returns **unless the data type has conflicts**, e.g.

```go
GetLocation() (float64, float64, error)
```
should be 
```go
GetLocation() (lat, long float64, err error)
```
but
```go
GetLatitude() (lat float64, err error)
```
should be 
```go
GetLatitude() (float64, error)
```

File: app/main.go [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package main
+
+import "github.com/ngavinsir/golangtraining/app/cmd"
+
+func main() {
+	cmd.Execute()
```
Comment (owenyuwono) [inline app/main.go:6]: cmd.Execute pattern is specific only to `spf13/cobra` library usage, if you don't use the library, you don't have to use this pattern

File: internal/postgres/paymentcodes.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	"github.com/google/uuid"
+	"github.com/ngavinsir/golangtraining"
+	"github.com/pkg/errors"
+)
+
+type paymentCodesRepository struct {
+	DB *sql.DB
+}
+
+func NewPaymentCodesRepository(db *sql.DB) golangtraining.IPaymentCodesRepository {
```
Comment (owenyuwono) [inline internal/postgres/paymentcodes.go]: let's practice "accept interface, return struct"
and declare interface at callsite

File: internal/postgres/paymentcodes.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+	if s.ExpirationDate.IsZero() {
+		s.ExpirationDate = time.Now().AddDate(51, 0, 0).UTC()
+	}
+
+	sqlStatement := `
+		INSERT INTO payment_code (id, payment_code, name, status, expiration_date, created_at, updated_at)
+		VALUES ($1, $2, $3, $4, $5, $6, $7)
+	`
+	_, err = p.DB.Exec(sqlStatement, s.ID, s.PaymentCode, s.Name, s.Status, s.ExpirationDate, s.CreatedAt, s.UpdatedAt)
```
Comment (owenyuwono) [inline internal/postgres/paymentcodes.go]: use ExecContext instead, so we can pass the context and set timeout or tracing for the db query

File: internal/postgres/paymentcodes.go [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+	DB *sql.DB
+}
+
+func NewPaymentCodesRepository(db *sql.DB) golangtraining.IPaymentCodesRepository {
+	return &paymentCodesRepository{
+		DB: db,
+	}
+}
+
+func (p paymentCodesRepository) Create(ctx context.Context, s *golangtraining.PaymentCode) (err error) {
```
Comment (owenyuwono) [inline internal/postgres/paymentcodes.go]:
```suggestion
func (p paymentCodesRepository) Create(ctx context.Context, s *golangtraining.PaymentCode) error {
```

File: internal/postgres/paymentcodes.go [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+		err = errors.Wrap(err, "cannot insert new payment code into DB")
+		return
+	}
+
+	return
+}
+
+func (p paymentCodesRepository) GetByID(ctx context.Context, ID string) (res golangtraining.PaymentCode, err error) {
+	sqlStatement := `SELECT * FROM payment_code where id=$1`
+	row := p.DB.QueryRow(sqlStatement, ID)
```
Comment (owenyuwono) [inline internal/postgres/paymentcodes.go]:
```suggestion
	row := p.DB.QueryRowContext(ctx, sqlStatement, ID)
```

File: internal/postgres/paymentcodes.go [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+	_, err = p.DB.Exec(sqlStatement, s.ID, s.PaymentCode, s.Name, s.Status, s.ExpirationDate, s.CreatedAt, s.UpdatedAt)
+	if err != nil {
+		err = errors.Wrap(err, "cannot insert new payment code into DB")
+		return
+	}
+
+	return
+}
+
+func (p paymentCodesRepository) GetByID(ctx context.Context, ID string) (res golangtraining.PaymentCode, err error) {
```
Comment (owenyuwono) [inline internal/postgres/paymentcodes.go]:
```suggestion
func (p paymentCodesRepository) GetByID(ctx context.Context, ID string) (golangtraining.PaymentCode, error) {
```

File: internal/postgres/paymentcodes.go [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+	sqlStatement := `SELECT * FROM payment_code where id=$1`
+	row := p.DB.QueryRow(sqlStatement, ID)
+	if err = row.Scan(
+		&res.ID, &res.PaymentCode, &res.Name, &res.Status,
+		&res.ExpirationDate, &res.CreatedAt, &res.UpdatedAt,
+	); err != nil {
+		err = errors.Wrap(err, "cannot get payment code by ID from DB")
```
Comment (owenyuwono) [inline internal/postgres/paymentcodes.go]: this package was intended for golang versions older than 1.13, it's included (although little bit modified) in the 1.13
```suggestion
		err = fmt.Errorf("%s: %w","cannot get payment code by ID from DB", err)
```
source: https://blog.golang.org/go1.13-errors

File: internal/rest/paymentcodes.go [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+	"io/ioutil"
+	"net/http"
+
+	"github.com/julienschmidt/httprouter"
+	golangtraining "github.com/ngavinsir/golangtraining"
+)
+
+type paymentCodesHandler struct {
+	repo golangtraining.IPaymentCodesRepository
+}
```
Comment (owenyuwono) [inline internal/rest/paymentcodes.go]:
```suggestion
type repository interface {
 	Create(ctx context.Context, p *golangtraining.PaymentCode) error
 	GetByID(ctx context.Context, ID string) (golangtraining.PaymentCode, error)
 }
type paymentCodesHandler struct {
	repo repository
}
func InitPaymentCodesHandler(r *httprouter.Router, repo repository) {
 	h := paymentCodesHandler{
 		repo: repo,
 	}

 	r.HandlerFunc("POST", "/payment-codes", h.create)
 	r.HandlerFunc("GET", "/payment-codes/:id", h.get)
 }
```

File: paymentcodes.go [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+	Status         string    `json:"status"`
+	ExpirationDate time.Time `json:"expiration_date"`
+	CreatedAt      time.Time `json:"created_at"`
+	UpdatedAt      time.Time `json:"updated_at"`
+}
+
+type IPaymentCodesRepository interface {
+	Create(ctx context.Context, p *PaymentCode) error
+	GetByID(ctx context.Context, ID string) (PaymentCode, error)
+}
```
Comment (owenyuwono) [inline paymentcodes.go]:
```suggestion
```

File: internal/rest/paymentcodes.go [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+
+func InitPaymentCodesHandler(r *httprouter.Router, repo golangtraining.IPaymentCodesRepository) {
+	h := paymentCodesHandler{
+		repo: repo,
+	}
+
+	r.HandlerFunc("POST", "/payment-codes", h.create)
+	r.HandlerFunc("GET", "/payment-codes/:id", h.get)
```
Comment (owenyuwono) [inline internal/rest/paymentcodes.go]:
```suggestion
	r.HandlerFunc(http.MethodPost, "/payment-codes", h.create)
	r.HandlerFunc(http.MethodGet, "/payment-codes/:id", h.get)
```

File: internal/rest/paymentcodes.go [hunk 12]
```diff
@@ -0,0 +1,62 @@
+package rest
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/julienschmidt/httprouter"
+	golangtraining "github.com/ngavinsir/golangtraining"
```
Comment (owenyuwono) [inline internal/rest/paymentcodes.go]:
```suggestion
	"github.com/ngavinsir/golangtraining"
```
(nitpick) unnecessary rename

File: app/cmd/root.go (hunk 1 above)
Comment (ngavinsir) [reply app/cmd/root.go:29]: ohh okay, TIL, thanks

**Comment by ngavinsir** [comment]
thanks bro @owenyuwono for the detailed review, learn a lot, will update the code

**Pull request merged**